lib/zlib.js

```diff
@@ -363,6 +363,9 @@ function Zlib(opts, mode) {
     var error = new Error(message);
     error.errno = errno;
     error.code = exports.codes[errno];
+    // TODO: Can't defer the error emission to nextTick because both the sync
```

Comment:
You can create two onerror handlers, one for the sync code and one for the async code:

The code in https://github.com/nodejs/node/blob/0b3936b49f06743ea7aa48f4c03c99edf48afe80/src/node_zlib.cc#L344-L367 would need to be duplicated (and shared through utility functions) for the sync and async versions - currently they both use `Zlib::Error` - they'd have to use `::Error` and `::ErrorAsync` and places that call `::Error` would have to be modified.

I think it would be a shame to block this PR on that though and it would require a lot more extensive review.